Add test global install path is used when DOTNET_ROOT does not exist

mateoatr opened the pull request:
The global installation path should be used when `DOTNET_ROOT` is specified but the path it points to does not exist.

Commit:
Add test

elinor-fung commented on src/installer/tests/HostActivation.Tests/MultiArchInstallLocation.cs:
```diff
@@ -71,6 +71,24 @@ public void EnvironmentVariable_ArchSpecificDotnetRootIsUsedOverDotnetRoot()
+                .DotNetRoot("non_existent_path")
+                .EnvironmentVariable("DOTNET_MULTILEVEL_LOOKUP", "0")
+                .Execute()
+                .Should().Pass()
```
This will only pass if there is an existing global installation, which we don't (and shouldn't) require for building the repo and running tests.

You should be able to also add the registered install location override and/or set the environment variable for overriding the default install path. I think that would cover this case where the host should fall back to those.

Commit:
PR feedback
Add test covering case where DOTNET_ROOT exists but does not have a dotnet installation on it

elinor-fung commented on src/installer/tests/HostActivation.Tests/MultiArchInstallLocation.cs:
```diff
@@ -71,6 +71,49 @@ public void EnvironmentVariable_ArchSpecificDotnetRootIsUsedOverDotnetRoot()
+                .EnvironmentVariable("DOTNET_MULTILEVEL_LOOKUP", "0")
+                .EnvironmentVariable(
+                    Constants.TestOnlyEnvironmentVariables.DefaultInstallPath,
+                    sharedTestState.InstallLocation)
```
You'll also need to configure the app to actually check the test-only environment variable - there's the `TestOnlyProductBehavior` class, which is also used by `RegisteredInstallLocationOverride`.

Commit:
Update tests

Commit:
Use product behavior for test where DOTNET_ROOT exists but has no host

Commit:
Fix test

elinor-fung commented on src/installer/tests/HostActivation.Tests/MultiArchInstallLocation.cs:
```diff
@@ -71,6 +71,56 @@ public void EnvironmentVariable_ArchSpecificDotnetRootIsUsedOverDotnetRoot()
+                        sharedTestState.InstallLocation)
+                    .Execute()
+                    .Should().Pass()
+                    .And.HaveStdErrContaining("Did not find [DOTNET_ROOT] directory [non_existent_path]")
```
Can this use the `HaveUsedDotNetRootInstallLocation` extension?

elinor-fung commented on src/installer/tests/HostActivation.Tests/MultiArchInstallLocation.cs:
```diff
@@ -71,6 +71,56 @@ public void EnvironmentVariable_ArchSpecificDotnetRootIsUsedOverDotnetRoot()
+                .Should().Fail()
+                .And.HaveStdErrContaining($"Using environment variable DOTNET_ROOT=[{projDir}]")
+                // If DOTNET_ROOT points to a folder that exists we assume that there's a dotnet installation in it
+                .And.HaveStdErrContaining("A fatal error occurred. The required library hostfxr.dll could not be found.");
```
> hostfxr.dll

There is `RuntimeInformationExtensions.GetSharedLibraryFileNameForCurrentPlatform` to get the platform-specific name.

Commit:
Get platform specific extension for hostfxr
Use HaveUsedNetRootInstallLocation
